File: sdk/prady-sdk/src/task.ts
import type { ScheduleOptions, ScheduledTask } from "./types.js";

async function readJson<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const detail = await response.text();
    throw new Error(detail || `HTTP ${response.status}`);
  }
  return (await response.json()) as T;
}

export class KryosTask {
  constructor(private readonly apiBase: string = "http://localhost:8005") {}

  async schedule(description: string, runAt: Date, options: ScheduleOptions = {}): Promise<string> {
    const response = await fetch(`${this.apiBase}/tasks/schedule`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ description, run_at: runAt.toISOString(), ...options }),
    });
    const data = await readJson<{ schedule_id: string }>(response);
    return data.schedule_id;
  }

  async cancel(scheduleId: string): Promise<void> {
    const response = await fetch(`${this.apiBase}/tasks/${encodeURIComponent(scheduleId)}`, {
      method: "DELETE",
    });
    await readJson<{ cancelled: boolean }>(response);
  }

  async list(): Promise<ScheduledTask[]> {
    const response = await fetch(`${this.apiBase}/tasks`);
    const data = await readJson<{ tasks?: ScheduledTask[] } | ScheduledTask[]>(response);
    return Array.isArray(data) ? data : data.tasks ?? [];
  }
}

File: sdk/prady-sdk-python/kryos_sdk/task.py
from __future__ import annotations

from dataclasses import dataclass

import httpx


@dataclass
class ScheduledTask:
    schedule_id: str
    description: str
    run_at: str
    status: str


class KryosTask:
    def __init__(self, api_base: str = "http://localhost:8005") -> None:
        self.api_base = api_base

    async def schedule(self, description: str, runAt, options: dict | None = None) -> str:
        payload = {"description": description, "run_at": runAt.isoformat()}
        if options:
            payload.update(options)
        async with httpx.AsyncClient() as client:
            resp = await client.post(f"{self.api_base}/tasks/schedule", json=payload)
        resp.raise_for_status()
        return resp.json()["schedule_id"]

    async def cancel(self, scheduleId: str) -> None:
        async with httpx.AsyncClient() as client:
            resp = await client.delete(f"{self.api_base}/tasks/{scheduleId}")
        resp.raise_for_status()

    async def list(self) -> list[ScheduledTask]:
        async with httpx.AsyncClient() as client:
            resp = await client.get(f"{self.api_base}/tasks")
        resp.raise_for_status()
        data = resp.json()
        items = data.get("tasks") if isinstance(data, dict) else data
        return [ScheduledTask(**item) for item in (items or [])]
